src/ros_can/src/node/node_ros_can.cpp

```diff
@@ -0,0 +1,214 @@
+#include "node/node_ros_can.hpp"
+
+#include <canlib.h>
+
+#include <chrono>
+#include <functional>
+#include <memory>
+#include <string>
+
+#include "rclcpp/rclcpp.hpp"
+#include "std_msgs/msg/string.hpp"
+
+RosCan::RosCan() : Node("node_ros_can") {
+  operationalStatus =
+      this->create_publisher<custom_interfaces::msg::OperationalStatus>("operationalStatus", 10);
+  wheelRPM = this->create_publisher<custom_interfaces::msg::WheelRPM>("wheelRPM", 10);
+  imu = this->create_publisher<sensor_msgs::msg::Imu>("imu", 10);
+  steeringAngle =
+      this->create_publisher<custom_interfaces::msg::SteeringAngle>("steeringAngle", 10);
+  busStatus = this->create_subscription<std_msgs::msg::String>(
+      "busStatus", 10, std::bind(&RosCan::busStatus_callback, this, std::placeholders::_1));
+  timer =
+      this->create_wall_timer(std::chrono::milliseconds(500), std::bind(&RosCan::canSniffer, this));
+  // initialize the CAN library
+  canInitializeLibrary();
+  // A channel to a CAN circuit is opened. The channel depend on the hardware
+  hnd = canOpenChannel(1, canOPEN_CAN_FD);
+  // Setup CAN parameters for the channel
+  stat = canSetBusParams(hnd, canBITRATE_500K, 0, 0, 0, 0, 0);  // check these values later
+  if (stat != canOK) {
+    RCLCPP_ERROR(this->get_logger(), "Failed to setup CAN parameters");
+  }
+  // There are different types of controllers, this is the default
+  stat = canSetBusOutputControl(hnd, canDRIVER_NORMAL);
+  if (stat != canOK) {
+    RCLCPP_ERROR(this->get_logger(), "Failed to setup CAN controller");
+  }
+}
+
+/**
+ * @brief Function to turn ON and OFF the CAN BUS
+ */
+void RosCan::busStatus_callback(std_msgs::msg::String busStatus) {
+  if (busStatus.data == "ON") {
+    hnd = canOpenChannel(0, canOPEN_CAN_FD);
+    stat = canBusOn(hnd);
+  } else if (busStatus.data == "OFF") {
+    stat = canBusOff(hnd);
+    canClose(hnd);
+  }
+}
+
+/**
+ * @brief Function cyclically reads all CAN msg from buffer
+ */
+void RosCan::canSniffer() {
+  long id;
+  unsigned char msg[8];
+  unsigned int dlc;
+  unsigned int flag;
+  unsigned long time;
+
+  stat = canRead(hnd, &id, &msg, &dlc, &flag, &time);
+  while (stat == canOK) {
+    // Convert the CAN message to a ROS message and publish it
+    canInterpreter(id, msg, dlc, flag, time);
+
+    stat = canRead(hnd, &id, &msg, &dlc, &flag, &time);
+  }
+}
+
+void RosCan::canInterpreter(long id, unsigned char msg[8], unsigned int dlc, unsigned int flag,
+                            unsigned long time) {
+  switch (id) {
+    case MASTER_STATUS:
+      switch (msg[0]) {
+        case 0x31:  // Current AS State
```

Comment:
Maybe extract the second level switches to other functions, to improve readability